pkg/galley/galley.go

```diff
@@ -16,48 +16,118 @@ package galley
 
 import (
 	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"time"
 
 	"google.golang.org/grpc"
 	mcpapi "istio.io/api/mcp/v1alpha1"
 
-	"istio.io/istio/pkg/log"
+	"istio.io/pkg/log"
 
+	"github.com/maistra/ior/pkg/bootstrap"
 	"github.com/maistra/ior/pkg/route"
 	networking "istio.io/api/networking/v1alpha3"
 	_ "istio.io/istio/galley/pkg/metadata" // Import the resource package to pull in all proto types.
-	mcpclient "istio.io/istio/pkg/mcp/client"
+	"istio.io/istio/pkg/mcp/creds"
 	"istio.io/istio/pkg/mcp/sink"
 	"istio.io/istio/pkg/mcp/testing/monitoring"
 )
 
-// ConnectToGalley ...
-func ConnectToGalley(galleyAddr string) {
-	ctx := context.Background()
-	conn, err := grpc.DialContext(ctx, galleyAddr, grpc.WithInsecure())
+const (
+	requiredCertCheckFreq = 1 * time.Second
+)
+
+// Galley is responsible to interact with Galley server
+type Galley struct {
+	args *bootstrap.Args
+}
+
+// New returns a Galley instance or an error
+func New(args *bootstrap.Args) (*Galley, error) {
+	if err := args.Validate(); err != nil {
+		return nil, fmt.Errorf("error validating arguments: %v", err)
+	}
+
+	log.Infof("Started IOR with\n%v", args)
+
+	return &Galley{args: args}, nil
+}
+
+// Run connects to Galley and runs the main loop
+func (g *Galley) Run() error {
+
+	url, err := url.Parse(g.args.McpAddr)
+
+	if err != nil {
+		return fmt.Errorf("invalid MCP URL %s %v", g.args.McpAddr, err)
+	}
+
+	securityOption := grpc.WithInsecure()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	switch url.Scheme {
+	case "mcp":
+	case "mcps":
+		if g.args.CredentialOptions == nil {
+			return fmt.Errorf("no credentials specified with secure MCP scheme")
+		}
+
+		requiredFiles := []string{g.args.CredentialOptions.CertificateFile, g.args.CredentialOptions.KeyFile, g.args.CredentialOptions.CACertificateFile}
+		log.Infof("Secure MCP configured. Waiting for required certificate files to become available: %v", requiredFiles)
+		for len(requiredFiles) > 0 {
```

Comment:
this seems more complex than it needs to be. Why are we changing elements in the array rather than just looping through?

Reply:
As we discussed in person, this is a copy from mixer code, and it makes this piece of code to use fewer lines than a regular foo loop. I'm also not a big fan, but I believe it doesn't hurt.